File: BOJ_STUDY/WEEK_6/김동훈/2644.js
const filePath = process.platform === "linux" ? "/dev/stdin" : "input.txt";
let input = (require("fs").readFileSync(filePath).toString().trim().split("\n"));

const people = +input[0];
const [target1, target2] = input[1].split(" ").map(e => +e);
const node = +input[2];
input = input.splice(3);
const graph = Array.from({ length: people+1 }).map(_ => []);

for (let i = 0; i < input.length; i++) {
    const [from, to] = input[i].split(" ").map(e => +e);
    graph[from].push(to);
    graph[to].push(from);
}

console.log(graph)
const bfs = (graph, startNode, targetNode) => {
    const visited = [];
    let needVisit = [[startNode, 0]];
    while (needVisit.length !== 0) {
        const [node, cnt] = needVisit.shift();
        // console.log('node,cnt', node, cnt);
        // 탈출조건
        if (node === targetNode) return cnt;

        // 이미 방문한 노드라면 탐색하지 않는다.
        if (!visited.includes(node)) {
            // 방문 배열 체크
            visited.push(node);
            // node + cnt 2차원 배열 만들기
            let nodes = graph[node].map((e) => [e, cnt + 1]);
            // needVisit배열 원소랑 nodes 원소 합쳐서 needVisit에 담기
            needVisit = needVisit.concat(nodes);
            // console.log('needVisit', needVisit)
        }
    }

    return -1;
};
const ans = bfs(graph, target1, target2); // graph, 7 , 3
console.log(ans);

File: BOJ_STUDY/WEEK_6/조정우/2644.py
from collections import deque

def bfs(v):
    q = deque()
    q.append(v)
    visited[v] = True
    ans = 0
    while q:
        ans += 1
        for _ in range(len(q)):
            k = q.popleft()
            if k == b:
                return ans-1
            for e in arr[k]:
                if visited[e] == False:
                    visited[e] = True
                    q.append(e)

    return -1

n = int(input())
a, b = map(int, input().split(' '))
m = int(input())
arr = [[] for _ in range(n+1)]
for _ in range(m):
    x, y = map(int, input().split(' '))
    arr[x].append(y)
    arr[y].append(x)

visited = [False] * (n+1)
print(bfs(a))


# input
# 9
# 7 3
# 7
# 1 2
# 1 3
# 2 7
# 2 8
# 2 9
# 4 5
# 4 6

#output
# 3
